src/coreclr/jit/lclmorph.cpp

```diff
@@ -2426,3 +2426,562 @@ void Compiler::fgExposeLocalsInBitVec(BitVec_ValArg_T bitVec)
+            int opNum = 1;
+            for (; opNum <= user->AsHWIntrinsic()->GetOperandCount(); opNum++)
+            {
+                if (user->AsHWIntrinsic()->Op(opNum) == convertOp)
+                {
+                    break;
+                }
+            }
+            assert(opNum <= user->AsHWIntrinsic()->GetOperandCount());
+
+            // Fix up the type of the lcl
+            lclOp->gtType = convertOp->gtType;
+
+            // Remove the convert convertOp
+            convertOp->gtBashToNOP();
+            *use = lclOp;
+            m_compiler->fgSequenceLocals(stmt);
+
+#ifdef DEBUG
+            JITDUMP("Updated V%02d to be a mask (Removed conversion)\n", lclOp->GetLclNum());
+            if (m_compiler->verbose)
+            {
+                m_compiler->gtDispTree(*use);
+            }
+#endif
+
+            return fgWalkResult::WALK_ABORT;
+        }
+
+        // Look for:
+        //      user(LCL_VAR(lclNum)))
+        else if ((*use)->OperIs(GT_LCL_VAR) && ((*use)->gtType != TYP_MASK) &&
+                 ((*use)->AsLclVarCommon()->GetLclNum() == lclNum) && !user->OperIsConvertVectorToMask())
+        {
+            GenTreeLclVar* lclOp = (*use)->AsLclVar();
+
+            // Fix up the type of the lcl
+            var_types vectorType = lclOp->gtType;
+            lclOp->gtType        = TYP_MASK;
+
+            // Create a convert to mask node and insert it infront of the lcl.
+            // There is not enough information in the lcl to get simd types. Instead we reuse the cached simd
+            // types from the removed convert nodes.
+            assert(simdBaseJitType != CORINFO_TYPE_UNDEF);
+            *use = m_compiler->gtNewSimdCvtMaskToVectorNode(vectorType, lclOp, simdBaseJitType, simdSize);
+
+#ifdef DEBUG
+            JITDUMP("Updated V%02d to be a mask (Added conversion)\n", lclOp->GetLclNum());
+            if (m_compiler->verbose)
+            {
+                m_compiler->gtDispTree(*use);
+            }
+#endif
+
+            return fgWalkResult::WALK_ABORT;
+        }
+
+        return fgWalkResult::WALK_CONTINUE;
+    }
+
+
+private:
+    unsigned    lclNum;
+    Statement*  stmt;
+    CorInfoType simdBaseJitType;
+    unsigned    simdSize;
+};
+
+//-----------------------------------------------------------------------------
+// fgLCLMasksCheckLCLStore: For the given statement, if it is a local store,
+// then update the store weights in the table.
+//
+// Arguments:
+//     stmt - The statement.
+//     weightsTable - table to update.
+//
+// Returns:
+//     True if a converted local store was found.
+//
+bool Compiler::fgLCLMasksCheckLCLStore(Statement* stmt, BasicBlock* const block, LCLMasksWeightTable* weightsTable)
+{
+    // Look for:
+    //      STORELCL(TYP_SIMD, ConvertMaskToVector(mask))
+
+    GenTree* tree = stmt->GetRootNode();
+
+    if ((tree->OperGet() != GT_STORE_LCL_VAR) || (!varTypeIsSIMD(tree->gtType)))
+    {
+        return false;
+    }
```

Comment:
There is no guarantee that the store is at the root of the statement. It can be deeper inside the tree due to embedded statements (`GT_COMMA`).

In some of this code it seems like the transformation is making an assumption that there is only one mask conversion in each statement, which isn't necessarily true. I would do all the accounting and conversions from inside some `GenTreeVisitor` implementations, and use `LocalsTreeList` purely as a way to figure out if running the visitors over some statement will be needed.

For example, this is what the similar "pass 1" in physical promotion looks like:
https://github.com/dotnet/runtime/blob/36bcc2c96045c6793dcfe3151c51a0597537917a/src/coreclr/jit/promotion.cpp#L2757-L2774

Reply:
... and reworked to have a single loop to gather the counts, then a single loop to do the converts, both using `GenTreeVisitor`, as suggested.